File: pa2/List.c
/*
  Name:Thomas Pham
  CruzID: 1593385
  Pa2
*/

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

//Structs -------------------------------------

//private NodeObj typedef
typedef struct NodeObj{
  int data;
  struct NodeObj* next;
  struct NodeObj* prev;
} NodeObj;

//private Node typedef
typedef NodeObj* Node;

//private ListObj type
typedef struct ListObj {
  Node front;
  Node back;
  Node cursor;
  int index;
  int length;
} ListObj;

//Constructors-Destructors ---------------------

//newNode()
//Returns refrence to new Node Object. Initializes next and
//data fields.
//Private.
Node newNode(int data){
  Node newNode = malloc(sizeof(NodeObj));
  newNode->data = data;
  newNode->next = NULL;
  return(newNode);
}

//freeNode()
//Frees heap memory pointed to by *pN, sets *pN to NULL.
//Private.
void freeNode(Node* pN){
  if( pN != NULL && *pN != NULL){
    free(*pN);
    *pN = NULL;
  }
}

//newList()
//Returns refrence to new empty list object
List newList(void){
  List L;
  L = malloc(sizeof(ListObj));
  L->front = L->back = L->cursor = NULL;
  L->length = 0;
  L->index = -1;
  return(L);
}

//freeList()
//Frees all heap memory associated with list *pL and sets
// *pL to NULL.
void freeList(List* pL){
  if(pL != NULL && *pL != NULL){
    Node currNode = (*pL)->front;
    while(currNode != NULL){
      deleteFront(*pL);
      currNode = (*pL)->front;
    }
    free(currNode);
  }
  free(*pL);
  *pL = NULL;
}

//Access Functions --------------------------------------

//Returns the length of the List
int length(List L){
  if(L == NULL){
    printf("List Error: calling length() on NULL List Ref");
    exit(1);
  }
  return(L->length);
}

//Returns the index of the cursor element
int index(List L){
  if(L == NULL){
    printf("List Error: calling index() on NULL List Ref");
    exit(1);
  }
  return(L->index);
}

//returns the front elements data.
int front(List L){
  if(L == NULL){
    printf("List Error: calling front() on NULL List Ref");
    exit(1);
  }
  return(L->front->data);
}

//Returns the back element's data
int back(List L){
  if(L == NULL){
    printf("List Error: calling back() on NULL List Ref");
    exit(1);
  }
  return(L->back->data);
}

//Returns the cursor's data
int get(List L){
  if(L == NULL || L->cursor == NULL){
    printf("List Error: calling get() on NULL List Ref or List->cursor is NULL");
    exit(1);
  }
  return(L->cursor->data);
}

//Returns 1 if and only if A contains the same integer sequence as B.
//otherwise return 0.
int equals(List A, List B){
  if(A == NULL || B == NULL){
    printf("List Error: Calling equals() on NULL List Ref");
    exit(1);
  }

  if(length(A) != length(B))
    return(0);

  Node aNode = A->front;
  Node bNode = B->front;

  while(aNode != NULL){
    if(aNode->data != bNode->data){
      aNode = bNode = NULL;
      return(0);
    }
    else {
      aNode = aNode->next;
      bNode = bNode->next;
    }
  }
  return(1);
}

//Manipulation procedures ------------------------
// Resets List to its original empty state.
void clear(List L){
  if(L == NULL){
    printf("List Error: Caling clear() on NULL List Ref");
    exit(1);
  }
  Node currNode = L->front;
  while(currNode != NULL){
    deleteFront(L);
    currNode = L->front;
  }
  freeNode(&currNode);
  L->front = L->back = L->cursor = NULL;
  L->length = 0;
  L->index = -1;
}

void moveFront(List L){// If List is non-empty, places the cursor under the front element,
  // otherwise does nothing.
  if(L == NULL){
    printf("List Error: Calling moveFront() on NULL List ref");
    exit(1);
  }
  if(L->length != 0){ //if length is not 0.
    L->cursor = L->front;
    L->index = 0;
  }
}

void moveBack(List L){// If List is non-empty, places the cursor under the back element,
 // otherwise does nothing.
  if(L == NULL){
    printf("List Error: Calling moveBack() on NULL List ref");
    exit(1);
  }
  if(L->length != 0){
    L->cursor = L->back;
    L->index = L->length - 1;
  }
}

void movePrev(List L){// If cursor is defined and not at front, moves cursor one step toward
 // front of this List, if cursor is defined and at front, cursor becomes
// undefined, if cursor is undefined does nothing.
  if(L == NULL){
    printf("List Error: Calling movePrev() on NULL List ref");
    exit(1);
  }
  if(L->cursor != NULL && L->index != 0){//somewhere in the middle
    L->cursor = L->cursor->prev;
    L->index--;
  }else if(L->cursor != NULL && L->index == 0){//at the front
    L->cursor = NULL;
    L->index = -1;
  }
}

void moveNext(List L){// If cursor is defined and not at back, moves cursor one step toward
 // back of this List, if cursor is defined and at back, cursor becomes
 // undefined, if cursor is undefined does nothing.
  if(L == NULL){
    printf("List Error: Calling moveNext() on NULL List ref");
    exit(1);
  }
  if(L->cursor != NULL && L->index != L->length - 1){//if cursor exits and not at the end
    L->cursor = L->cursor->next;
    L->index++;
  }else if(L->cursor != NULL && L->index == L->length - 1){//if cursor exists and AT the end
    L->index = -1;
    L->cursor = NULL ;
  }else if(L->cursor == NULL){} //do nothing
}

void prepend(List L,int data){
  // Insert new element into this List. If List is non-empty,
  // insertion takes place before front element.
  if(L == NULL){
   printf("List Error: Calling prepend() on NULL List ref");
   exit(1);
  }
  Node nNode = newNode(data);
  nNode->prev = NULL;
  nNode->next = L->front;
  if(L->length > 0)
    L->front->prev = nNode;
  L->front = nNode;
  L->length++;
  if(L->cursor != NULL)//if cursor is defined
    L->index++;
  if(L->length == 1)//only one node in the list
    L->back = nNode;
}

void append(List L, int data){// Insert new element into this List. If List is non-empty,
 // insertion takes place after back element.
  if(L == NULL){
    printf("List Error: Calling append() on NULL List ref");
    exit(1);
  }
  Node nNode = newNode(data);

  if(L->length == 0){//if nothing in list
    L->front = nNode;
    L->back = nNode;
    L->length++;
  }else{
    L->back->next = nNode;
    nNode->prev = L->back;
    L->back = nNode;
    L->length++;
  }
}

void insertBefore(List L, int data){//insert before cursor
  if(L == NULL || L->cursor == NULL){
    printf("List Error: Calling insertBefore() on NULL List ref OR L->cursor");
    exit(1);
  }
  if(L->index == 0)
    prepend(L,data);
  else {
    Node nNode = newNode(data);
    nNode->prev = L->cursor->prev;
    nNode->next = L->cursor;
    L->cursor->prev->next = nNode;
    L->cursor->prev = nNode;
    L->index++;
    L->length++;
  }
}

void insertAfter(List L, int data){//insert element after cursor
  if(L == NULL || L->cursor == NULL){
    printf("List Error: Calling insertAfter() on NULL List ref OR L->cursor");
    exit(1);
  }
  if(L->index == L->length - 1){//index is at the end of the list
    append(L,data);
  }else {
    Node nNode = newNode(data);
    L->cursor->next->prev = nNode;
    nNode->next = L->cursor->next;
    L->cursor->next = nNode;
    nNode->prev = L->cursor;
    L->length++;
  }
}

void deleteFront(List L){//delete Front of list.
  if(L == NULL){
    printf("List Error: Calling deleteFront() on NULL List ref");
    exit(1);
  }
  if(L->length == 0){
    printf("List Error: Calling deleteFront() on empty List");
    exit(1);
  }
  Node nNode = L->front;

  if(L->length > 1){
    L->front = L->front->next;
  } else{
    L->front = L->back = NULL;
  }
  freeNode(&nNode);
  L->length--;
  if(L->index == 0){//if cursor is at the beginning, set to undefined.
    L->index = -1;
    L->cursor = NULL;
  }
  if(L->cursor != NULL) // if cursor exists
    L->index--;
}

void deleteBack(List L){
  if(L == NULL){
    printf("List Error: calling deleteBack() on a NULL list\n");
    exit(1);
  }else if(L->length == 0){
    printf("List Error: calling deleteBack() on an empty list\n");
    exit(1);
  }else if(L->length == 1){//if just one element in list
    if(L->index == 0){ //if cursor is defined in a one elemnt list, delete cursor
      L->index = -1;
      L->cursor = NULL;
    }
    freeNode(&L->front);//deleting only element
    L->front = NULL;
    L->back = NULL;
    L->length = 0;
  }else{
    if(L->cursor == L->back){
      L->cursor = NULL;
      L->index = -1;
    }
    Node nNode = L->back;
    L->back = L->back->prev;
    L->back->next = NULL;
    freeNode(&nNode);
    L->length--;
  }
}

void delete(List L){
  if(L == NULL || L->cursor == NULL){
    printf("List Error: Calling delete() on NULL List ref OR L->cursor");
    exit(1);
  }

  if(L->index == 0)//cursor is at the first element
    deleteFront(L);
  else{
    Node oldNode = L->cursor;
    oldNode->prev->next = oldNode->next;
    if(L->index == L->length - 1)//if cursor is last element
      L->back = oldNode->prev;
    else
      oldNode->next->prev = oldNode->prev;
    freeNode(&oldNode);
    L->cursor = NULL;
    L->length--;
    L->index = -1;
  }
}

void printList(FILE *out, List L){
  moveFront(L);
  while(L->cursor != NULL){
    fprintf(out, "%d ",get(L));
    moveNext(L);
  }
}

List copyList(List L){
  if(L == NULL){
    printf("List Error: Calling copyList() on NULL List ref");
    exit(1);
  }
  List nList = newList();
  Node currNode = L->front;
  while(currNode != NULL){
    append(nList,currNode->data);
    currNode = currNode->next;
  }
  return(nList);
}

File: pa3/List.java
/*
  Thomas Pham
  CruzID: 1593385
  pa1
*/
public class List {

  private class Node {
    // Fields..
    Object data;
    Node next;
    Node prev;

    // Constructor
    Node(Object data) {
      this.data = data;
      this.next = null;
      this.prev = null;
    }

    // Overrides Object's toString() method.
    public String toString() {
      return String.valueOf(data);
    }
  }

  /* Start of List Class */

  private int length;
  private Node front;// this is the first node in the list
  private Node cursor;
  private int index;
  private Node back; // this is the last node in the list
  // Constructor

  List() { // Creates a new empty list.
    this.length = 0;
    this.cursor = null;
    this.index = -1;
    this.front = null;
    this.back = null;
  }

  // Access functions
  public int length() { // Returns the number of elements in this List.
    return length;
  }

  public int index() { // If cursor is defined, returns the index of the cursor element,
    // otherwise returns -1.
    if (cursor != null) {
      return index;
    } else
      return -1;
  }

  public Object front() { // Returns front element. Pre: length()>0
    return front.data;
  }

  public Object back() { // Returns back element. Pre: length()>0
    return back.data;
  }

  public Object get() { // Returns cursor element. Pre: length()>0, index()>=0
    return cursor.data;
  }

  public boolean equals(List L) { // Returns true if and only if this List and L are the same
    // integer sequence. The states of the cursors in the two Lists
    // are not used in determining equality.
    if (length != L.length) // if the lengths are not the same, they are not equal lists.
      return false;

    Node l1Node = front;
    Node l2Node = L.front;
    while (l1Node != null) { // walking through the entire list
      if (l1Node.data != l2Node.data)
        return false;
      else {
        l1Node = l1Node.next;
        l2Node = l2Node.next;
      }
    }
    return true;
  }

  // Manipulation procedures

  public void clear() { // Resets this List to its original empty state.
    length = 0;
    cursor = null;
    index = -1;
    front = null;
    back = null;
  }

  public void moveFront() { // If List is non-empty, places the cursor under the front element,
    // otherwise does nothing.
    if (length != 0) {
      cursor = front;
      index = 0;
    }
  }

  public void moveBack() { // If List is non-empty, places the cursor under the back element,
    // otherwise does nothing.
    if (length != 0) {
      cursor = back;
      index = length - 1;
    }
  }

  public void movePrev() { // If cursor is defined and not at front, moves cursor one step toward
    // front of this List, if cursor is defined and at front, cursor becomes
    // undefined, if cursor is undefined does nothing.
    if (cursor != null && index != 0) {
      cursor = cursor.prev;
      index--;
    } else if (cursor != null && index == 0) {
      cursor = null;
      index = -1;
    } else if (cursor == null) {
    } // do nothing
  }

  public void moveNext() { // If cursor is defined and not at back, moves cursor one step toward
    // back of this List, if cursor is defined and at back, cursor becomes
    // undefined, if cursor is undefined does nothing.
    if (cursor != null && index != length - 1) {
      cursor = cursor.next;
      index++;
    } else if (cursor != null && index == length - 1) {
      cursor = null;
      index = -1;
    } else if (cursor == null) {
    } // do nothing
  }

  public void prepend(Object data) { // Insert new element into this List. If List is non-empty,
    // insertion takes place before front element.
    Node newNode = new Node(data);
    newNode.prev = null;
    newNode.next = front;
    if (length > 0) // if not the only element in the list
      front.prev = newNode;
    front = newNode;
    length++;

    if (cursor != null) { // if the cursor is defined
      index++;
    }
    if (length == 1) { // if there is only one node in the list
      back = newNode;
    }

  }

  public void append(Object data) { // Insert new element into this List. If List is non-empty,
    // insertion takes place after back element.
    Node newNode = new Node(data);

    if (length == 0) { // if theres nothing in the list,
      prepend(data);
    } else {// adding to the front of the list
      back.next = newNode;
      newNode.prev = back;
      back = newNode;
      length++;
    }

  }

  public void insertBefore(Object data) { // Insert new element before cursor.
    // Pre: length()>0, index()>=0
    Node newNode = new Node(data);
    if (index == 0)
      prepend(data);
    else {
      newNode.prev = cursor.prev;
      newNode.next = cursor;
      cursor.prev.next = newNode;
      cursor.prev = newNode;
      index++;
      length++;
    }
  }

  public void insertAfter(Object data) { // Inserts new element after cursor.
    // Pre: length()>0, index()>=
    Node newNode = new Node(data);
    if (index == length - 1) { // if index is at the end of the list
      cursor.next = newNode;
      newNode.prev = cursor;
      back = newNode;
      length++;
    } else {
      cursor.next.prev = newNode;
      cursor.next = newNode;
      newNode.prev = cursor;
      length++;
    }
  }

  public void deleteFront() { // Deletes the front element. Pre: length()>0
    front = front.next;
    length--;

    if (index == 0) { // if cursor is at the beginning, set cursor to null
      index = -1;
      cursor = null;
    }
    if (cursor != null) {
      index--;
    }
  }

  public void deleteBack() { // Deletes the back element. Pre: length()>0
    back = back.prev;
    if (back != null) {
      back.next = null;
    } else { // only one element in the list
      front = null;
    }
    length--;
    if (index == length) { // if cursor is at the end, set cursor to null
      index = -1;
      cursor = null;
    }
  }

  public void delete() { // Deletes cursor element, making cursor undefined.
    // Pre: length()>0, index()>=0
    if (cursor == null) {
      // do nothing
    } else if (index == 0) {// cursor is the first element
      deleteFront();
    } else {
      Node prevNode = cursor.prev;
      prevNode.next = cursor.next;
      if (index == length - 1)// if cursor is last element;
        back = prevNode;
      cursor = null;
      length--;
      index = -1;
    }
  }

  public String toString() { // Overrides Object's toString method. Returns a String
    // representation of this List consisting of a space
    // separated sequence of integers, with front on left.
    if (length == 0)
      return new String("");

    Node currNode = front;
    String list = "";
    while (currNode != null) {
      list += currNode.data + " ";
      currNode = currNode.next;
    }
    list = list.substring(0, list.length() - 1);
    return list;
  }

}
